pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go

```diff
@@ -108,6 +110,8 @@ func mutateVMIs(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
 	// Apply default cpu model
 	setDefaultCPUModel(&vmi, informers.ConfigMapInformer.GetStore())
 
+	setDefaultCPUTime(&vmi)
```

Comment:
This needs to be applied after `applyNamespaceLimitRangeValues`, and you will have to adjust `applyNamespaceLimitRangeValues`, to set the default cpu time from the limitrange object. Otherwise you would always override limitranges set in a namespace.

Reply:
@rmohr , Oh, I think I see your point - you mean that if `applyNamespaceLimitRangeValues` is ever changed to set CPU defaultRequest when CPU-request is not already set then setting CPU-request in `setDefaultCPUTime` that is called beforehand would make the aforementioned logic redundant?

That seems to be the case with memory-request at the moment btw (kubevirt sets it to 8192Ki so a similar logic in `applyNamespaceLimitRangeValues` that sets memory-request to memory defaultRequest from limitranges would not work, right?).

I wonder though why to introduce defaults in kubevirt-config then and not to use limitranges - is it because we assume different default requests for VMs and ordinary pods but cannot assume different namespaces for VMs and ordinary pods? If that's the reason, is it possible to we add a LimitType of VMI?

Reply:
> That seems to be the case with memory-request at the moment btw (kubevirt sets it to 8192Ki so a similar logic in `applyNamespaceLimitRangeValues` that sets memory-request to memory defaultRequest from limitranges would not work, right?).

@ahadas good point, you are right that it would not happen for memory requests. I think that this is a bug. I think we should apply the limits and requests from `LimitRange` 1:1 already in virt-api, but without adding any virt specific defaults. The defaults or calculated overheads should probably stay in virt-controller (where the overhead calculation of memory resides). Defaults tend to become dynamic (like memory overhead) and virt-api does not necessarily have all information present which is needed to calculate that (for example, referenced PVCs and PVs, maps and secrets don't have to be present when creating VMIs or pods).

Reply:
Here another related issue: https://github.com/kubevirt/kubevirt/issues/1348

Reply:
@rmohr ,
1. I don't mind fixing it for the memory as well. But the question is how :) currently, the code finds the minimal default limit for the memory although that doesn't seem to be what k8s does when multiple limitranges are set for containers in the namespace (iiuc the results I see then it picks the last one that was set). However, taking the minimal limit does make sense. But for defaultRequest the logic should probably to take the one that is set in the limit we have taken the default limit from, or otherwise (i.e., when no default limit is taken) - take a random defaultRequest from one of the defined limits. What do you think?

2. About not having virt-specific defaults and having defaults dynamic - I don't know, it does make sense to me to provide different defaults for VMs compared to ordinary pods due to their different nature (run for longer time, being more idle) and computing the defaults seems to be more complex than what we planned. @jeniferh , @MarSik  - thoughts?

Reply:
> 2\. About not having virt-specific defaults and having dynamically calculated defaults - I don't know, it does make sense to me to provide different defaults for VMs compared to ordinary pods due to their different nature (run for longer time, being more idle) and computing the defaults seems to be more complex than what we planned. @jeniferh , @MarSik  - thoughts?

Yeah we can have virt defaults. I just meant that they should be calculated and added by virt-controller and not virt-api.

> 1. I don't mind fixing it for the memory as well. But the question is how :) currently, the code finds the minimal default limit for the memory although that doesn't seem to be what k8s does when multiple limitranges are set for containers in the namespace (iiuc the results I see then it picks the last one that was set). But taking the minimal limit does make sense. For defaultRequest, however, the logic should probably to take the one that is set in the limit we have taken the default limit from, or otherwise (i.e., when no default limit is taken) - take a random defaultRequest from one of the defined limits. What do you think?

Anything which is different to what k8s does is wrong here. If we diverge here we will have to change it.

Reply:
So I think it should be like this:

 1) In virt-api apply range limits on the `VMI` for all values the same way like k8s does it on Pods.
 2) In virt-controller add memory-guest-overhead if appropriate, add cpu-requests if appropriate on the `Pod`.

Reply:
I am however not entirely sure here, since we will always have this small time-gap between posting the VMI and creating the actual pods, so there could be a range limit which got active exactly in that small window.

Reply:
> I am however not entirely sure here, since we will always have this small time-gap between posting the VMI and creating the actual pods, so there could be a range limit which got active exactly in that small window.

Right, but I don't think it is that important,
- if the VMI is posted after the range limit was set, it will get the up-to-date limits
- if the VMI is posted before the range limit was set then we may not get the most up-to-date limits but such races would always be there I suppose (even if we had fetched the limits and then set them directly on the pod, a new limit could have been set right after we queried the limits, right?)

So I don't think we can/should promise anything about entities that are received before a range limit is set

Reply:
@rmohr , posted #2150 to change namespace-limits as discussed - could you please review?

Reply:
Moved that code to api/v1/defaults.go so it would be executed after namespace-limits are applied - for that, I had to break the dependency between virt-config and kubecli packages.
The namespace-limits and memory related stuff are handled separately by other PRs.
@rmohr , could you please do a follow-up review?